### PR TITLE
Move UEFI to special category in Custom spoke (#1183880)

### PR DESCRIPTION
**Pull request by jkonecny12**
The UEFI partitions are moved to special category to prevent user from
unintentional deletion which could cause existing systems to be
unbootable. This category is used only if the UEFI is shared between
installed systems or if there are any unknown filesystems (could be
Windows UEFI).

*Resolves: rhbz#1183880*

--------------------
Screenshots:
### UEFI shared
UEFI shared

### UEFI doesn't show up in the existing system
UEFI doesn't show up in the existing system

### Removed RHEL system by "Delete all" button
UEFI removed system by "Delete all" button

### Move all UEFIs when any unknown partition present
Move all UEFIs when any unknown partition present

### UEFI without sharing between systems and without unknown partition
UEFI without sharing

### CODE REVIEW

**Comment by vpodzime** [comment]
Looks good to me and I like the approach to the (real) issue.

**Comment by clumens** [comment]
I have several general, nitpicky comments:

* I don't like the name "UEFI partitions".  We've avoided the word "partition" in general, and as adamw pointed out in IRC is seems awfully specific.  We'll likely have this same situation on other platforms.  Perhaps the name should be more generic.

* I think it's accidental that these are showing up under a DATA heading.  Regardless, I don't think it belongs under any heading.  I think there should probably be a different Page subclass that these guys live in, similar to how there's an Unknown page.  In fact you might just be able to borrow from that.

* I find it a little weird that there can be multiple selectors with different names but the same device.  I assume that's because it's part of multiple OSes?  I feel like that's hard to discover here.  Could they perhaps be grouped together but separated from other possibly shared UEFI partitions?  That is, in the next-to-last screenshot, could the two sda1s be their own page, vda1s be their own, etc.?  I don't know if that would look good though.  It might be more confusing.

* What happens when you delete just the first sda1 selector in the next-to-last screenshot?  Does the other one disappear as well, or do you have to click delete twice to actually delete one thing?

Thanks for taking a look at this, though.  UI stuff is just always tons of nitpicky details.

**Comment by AdamWill** [comment]
@clumens note I wasn't just talking about the UI - I was thinking about the way the actual code is written. This code is very specifically tailored to UEFI, I was thinking it might make sense for it to be a bit more generic and use the data that's in the platform stuff about shared boot partitions and stuff (as a rough thought, I haven't looked into it in any detail).

File: pyanaconda/ui/gui/spokes/custom.py [hunk 1]
```diff
@@ -500,6 +500,7 @@ def _populate_accordion(self):
         log.debug("ui: new_devices=%s", [d.name for d in new_devices])
 
         ui_roots = self._storage_playground.roots[:]
+        page_efi = Page(_("UEFI partitions"))
```
Comment (AdamWill) [inline pyanaconda/ui/gui/spokes/custom.py]: Perhaps this could be `UnknownPage`, as we don't need the SYSTEM/DATA sorting stuff?

**Comment by AdamWill** [comment]
How do folks feel about this possible alternative?

    diff --git a/pyanaconda/ui/gui/spokes/custom.py b/pyanaconda/ui/gui/spokes/custom.py
    index 24c8474..4405097 100644
    --- a/pyanaconda/ui/gui/spokes/custom.py
    +++ b/pyanaconda/ui/gui/spokes/custom.py
    @@ -69,6 +69,7 @@ from blivet.errors import NoDisksError
     from blivet.errors import NotEnoughFreeSpaceError
     from blivet.devicelibs import raid, crypto
     from blivet.devices import LUKSDevice
    +from blivet.platform import platform
     
     from pyanaconda.storage_utils import ui_storage_logger, device_type_from_autopart
     from pyanaconda.storage_utils import DEVICE_TEXT_PARTITION, DEVICE_TEXT_MAP, DEVICE_TEXT_MD
    @@ -1962,10 +1963,28 @@ class CustomPartitioningSpoke(NormalSpoke, StorageChecker):
                         return
     
                 if dialog.deleteAll:
    -                for dev in (s._device for s in page.members):
    -                    self._destroy_device(dev)
    -            else:
    -                self._destroy_device(device)
    +                protected = platform.bootStage1ConstraintDict["format_types"]
    +                otherpgs = (pg for pg in self._accordion.allPages if
    +                            pg.pageTitle != page.pageTitle)
    +                otherdevs = []
    +                log.info("deleteall: otherdevs: %s", otherdevs)
    +                for otherpg in otherpgs:
    +                    otherdevs.append(mem._device.id for mem in otherpg.members)
    +                # we never want to delete known-shared devs here. we also
    +                # exclude 'device' as we'll unconditionally delete it later
    +                for dev in (s._device for s in page.members if
    +                            s._device.id not in otherdevs and
    +                            s._device.id != device.id):
    +                    # we only want to delete boot partitions if they're not
    +                    # shared *and* we have no unknown partitions
    +                    if (not self.unusedDevices or dev.format.type not in protected):
    +                        log.info("deleteall: removed %s", dev.name)
    +                        self._destroy_device(dev)
    +                    else:
    +                        log.info("deleteall: didn't remove %s", dev.name)
    +
    +            # We do this even in deleteAll case, see above comment
    +            self._destroy_device(device)
     
             log.info("ui: removed device %s", device.name)
     
Basically, when 'delete all' is clicked, don't delete any partitions that are 'shared' (appear in other roots), or bootloader partitions if there are any unknown partitions (stole that idea from @jkonecny12 for the same reason - we don't detect all OSes, so we use this as a proxy). I've tested it works in a simple UEFI case, haven't tested the `otherdevs` (shared partitions) stuff yet. I can test this and work on it some more if people are interested in the approach. If we do this we *could* adjust the 'delete all' text to explain more clearly what will happen, and maybe the delete confirm dialog text for bootloader partitions, but of course those are string freeze breaks.

**Comment by jkonecny12** [comment]
> I don't like the name "UEFI partitions". We've avoided the word "partition" in general, and as adamw pointed out in IRC is seems awfully specific. We'll likely have this same situation on other platforms. Perhaps the name should be more generic.

Are you have some idea about how it should be named? Only "UEFI" seems little odd.

> I think it's accidental that these are showing up under a DATA heading. Regardless, I don't think it belongs under any heading. I think there should probably be a different Page subclass that these guys live in, similar to how there's an Unknown page. In fact you might just be able to borrow from that.
 
I missed this solution, good idea. I tested it and it works nice. I'll change it to `UnknownPage`.

> I find it a little weird that there can be multiple selectors with different names but the same device. I assume that's because it's part of multiple OSes? I feel like that's hard to discover here. Could they perhaps be grouped together but separated from other possibly shared UEFI partitions? That is, in the next-to-last screenshot, could the two sda1s be their own page, vda1s be their own, etc.? I don't know if that would look good though. It might be more confusing.

Yeah that could be but how you want to name the category `UEFI Shared` any idea? Also this will make the patch bigger. I think this part should go only to master. Question is, if only this part or whole patch and use the @AdamWill version instead.

> What happens when you delete just the first sda1 selector in the next-to-last screenshot? Does the other one disappear as well, or do you have to click delete twice to actually delete one thing?

Now it remove all sda1 items. I personally don't know which method is better if remove all or only one? From implementation point of view there could be some counter but as before it makes the patch bigger and it should go only to master though.

File: pyanaconda/ui/gui/spokes/custom.py (hunk 1 above)
Comment (jkonecny12) [inline pyanaconda/ui/gui/spokes/custom.py]: I already tested it and it works nice. I'm going to change it to `UnknownPage`. Thank you for advice.

**Comment by dwlehman** [comment]
The checkbox says "Delete all other file systems in the ___ root as well". We should not start doing something different unless we're also going to change that string to describe the silliness we're trying to accomplish here.

**Comment by jkonecny12** [comment]
@AdamWill are you think that my patch is so invasive for F23? It's only one method change.

**Comment by AdamWill** [comment]
@dwlehman right, as I said, a full version of my alternative would go along with a change to that string and also possibly a change to the delete confirmation dialog text for bootloader partitions. I just wanted to float the *idea* first, though, so I didn't waste time working on it further if no-one was interested.

@jkonecny12 Well, I'm not sure, I just wanted to provide a possible alternative. I'm a *little* worried about yours because it's a fairly significant UI change. For instance, the 'roots' are really supposed to be operating systems; now you're inventing one that isn't. We probably have docs that talk about these groups (has anyone looked at the 'Help' text?) - they'd have to change. The issues @clumens noted with the same ESP appearing multiple times in the 'UEFI partitions' root seem a bit difficult to deal with cleanly. And another consequence I thought of: when you delete a partition in the 'UEFI partitions' root you get that same 'do you want to delete all the other partitions in this root?' question, which really doesn't make a lot of sense.

We should probably ask mizmo about this?

**Comment by AdamWill** [comment]
As for the naming question, I would call it something like 'System boot partitions'.

**Comment by dwlehman** [comment]
I don't really think there is a legitimate problem to fix here, but if we are to address it, I would prefer we do it quietly, like @AdamWill is proposing. There's no need for more UI here aside from updating strings in existing dialog(s).

**Comment by dwlehman** [comment]
```patch
diff --git a/pyanaconda/ui/gui/spokes/lib/custom_storage_helpers.py b/pyanaconda/ui/gui/spokes/lib/custom_storage_helpers.py
index e095fee..a3d7aa9 100644
--- a/pyanaconda/ui/gui/spokes/lib/custom_storage_helpers.py
+++ b/pyanaconda/ui/gui/spokes/lib/custom_storage_helpers.py
@@ -355,7 +355,9 @@ class ConfirmDeleteDialog(GUIObject):
             rootName = rootName.replace("_", "__")
         self._removeAll.set_label(
                 C_("GUI|Custom Partitioning|Confirm Delete Dialog",
-                    "Delete _all other file systems in the %s root as well.")
+                    "Delete _all other file systems in the %s root as well, "
+                    "including those used by other installed operating "
+                    "systems.")
                 % rootName)
         self._removeAll.set_sensitive(rootName is not None)
 ```

This is about as much as is warranted in my opinion.

**Comment by AdamWill** [comment]
@dwlehman so current status is the bug was rejected as a blocker, so you can do whatever you like here, but it'd be nice if we could at least pick one thing (this, my #383 , or your text change) and do it? if that's just your text change, so be it, but even that is a definite improvement on current behaviour - at least it cues the user explicitly that the button might be dangerous and they should stop and think.

**Comment by jkonecny12** [comment]
There were better solutions so I'm closing this PR.
